base/special/expm1.jl

```diff
@@ -0,0 +1,232 @@
+expm1_huge(::Type{Float32}) = 18.714973f0 # around 27*ln2, taken from source
+expm1_overflow(::Type{Float32}) = 88.72168f0
+expm1_underflow(::Type{Float32}) = 2.9802322f-8 # 2f0^-25
+
+@inline expm1_ln2(T) = expm1_ln2_hi(T) + expm1_ln2_lo(T)
+
+expm1_big_k(::Type{Float64}) = 1024
+expm1_big_k(::Type{Float32}) = 128
+expm1_small_k(::Type{Float64}) = Int32(20)
+expm1_small_k(::Type{Float32}) = Int32(23)
+
+# Domain [-0.34568, 0.34568], range ~[-6.694e-10, 6.696e-10]:
+# |6 / x * (1 + 2 * (1 / (exp(x) - 1) - 1 / x)) - q(x)| < 2**-30.04
+# Scaled coefficients: Qn_here = 2**n * Qn_for_q (see s_expm1.c):
+@inline expm1_p(hfxs::Float32) = @horner(hfxs, -3.3333212137f-2, # Q1
+                                        1.5807170421f-3) # Q2
+
+@inline expm1_two_pow_mk(::Type{Float32}, k) = reinterpret(Float32, UInt32((0x7f-k)<<23))
+@inline expm1_two_pow_mk(::Type{Float64}, k) = reinterpret(Float64, UInt64(0x3ff00000+((-k)<<20))<<32)
+
+@inline expm1_two_pow_k(::Type{Float32}, k) = reinterpret(Float32, UInt32(0x3f800000+(k<<23)))
+@inline expm1_two_pow_k(::Type{Float64}, k) = reinterpret(Float64, UInt64(0x3ff00000+(k<<20))<<32)
+
+#this is 2^-k not 1-2f0^k
+#@inline expm1_one_m_2_pow_mk(::Type{Float32}, k) = 
+@inline expm1_one_m_2_pow_mk(::Type{Float32}, k) = reinterpret(Float32, UInt32(0x3f800000 - (0x1000000>>k)))
+# 0x00000000 after & comes from lowword(1.0) = unsafe_trunc(UInt32, reinterpret(UInt64, 1.0))
+#@inline expm1_one_m_2_pow_mk(::Type{Float64}, k) = reinterpret(Float64, (UInt64(0x3ff00000 - (0x200000>>k))<<32)|0x00000000)
+@inline expm1_one_m_2_pow_mk(::Type{Float64}, k) = reinterpret(Float64, (UInt64(0x3ff00000 - (0x200000>>k))<<32) | unsafe_trunc(UInt32, reinterpret(UInt64, 1.0)))
+
+@inline expm1_two_pow_big_km1(T::Type{Float64}) = 8.98846567431158e307 # T(2.0)^(expm1_big_k(T) - 1)
+@inline expm1_two_pow_big_km1(T::Type{Float32}) = 1.7014118f38# T(2.0)^(expm1_big_k(T) - 1)
+
+function expm1(x::T) where T<:Union{Float32, Float64}
+    xsign = signbit(x)
+    absx = abs(x)
+ 
+    # filter out huge and non-finite argument
+	if absx >= expm1_huge(T) # quit early for large absolute values of x
+        if isinf(absx)
+            return xsign == true ? x : -T(1.0)
+        end
+        if x >= expm1_overflow(T) # largest value T can take before overflowing
+            return T(Inf)
+        end
+        if xsign == true # x < -expm1_huge(T)
+	    	return -T(1.0)
+        end
+    end
+    
+    # argument reduction
+	if absx > T(0.5)*expm1_ln2(T)
+	    if absx < T(1.5)*expm1_ln2(T)
+            if xsign == false
+                hi = x - expm1_ln2_hi(T)
+                lo = expm1_ln2_lo(T)
+                k = Int32(1)
+            else
+                hi = x + expm1_ln2_hi(T)
+                lo = -expm1_ln2_lo(T)
+                k = Int32(-1)
+            end
+	    else
+            #            k  = round(Int32, expm1_invln2(T)*x)
+            k = unsafe_trunc(Int32, expm1_invln2(T)*x + (!xsign ? 0.5 : -0.5))
```

Comment:
possibly `ifelse(xsign, -0.5, 0.5)`

Reply:
Also, `T(0.5)`?

Reply:
to avoid the branch? and yes, T(0.5) to be safe!